core/drivers/versal_trng.c

```diff
@@ -0,0 +1,1150 @@
+	memset(&trng->usr_cfg, 0, sizeof(trng->usr_cfg));
+	memset(trng->buf, 0, sizeof(trng->buf));
+	memset(trng->dfout, 0, sizeof(trng->dfout));
+	trng->status = TRNG_UNINITIALIZED;
+
+	return TEE_SUCCESS;
+error:
+	trng->status = TRNG_ERROR;
+
+	return TEE_ERROR_GENERIC;
+}
+
+/* Health tests should be run when the configured mode is of PTRNG or HRNG */
+static TEE_Result trng_health_test(struct versal_trng *trng)
+{
+	struct trng_usr_cfg tests = {
+		.mode = TRNG_HRNG,
+		.seed_life = 10,
+		.dfmul = 7,
+		.predict_en = false,
+		.iseed_en = false,
+		.pstr_en = false,
+		.df_disable = false,
+	};
+
+	if (trng_instantiate(trng, &tests))
+		goto error;
+
+	if (trng_release(trng))
+		goto error;
+
+	return TEE_SUCCESS;
+error:
+	trng->status = TRNG_ERROR;
+
+	return TEE_ERROR_GENERIC;
+}
+
+/*
+ * The KAT test should be run when the TRNG is configured in DRNG or HRNG mode.
+ * If KAT fails, the driver has to be put in error state.
+ */
+static TEE_Result trng_kat_test(struct versal_trng *trng)
+{
+	struct trng_usr_cfg tests = {
+		.mode = TRNG_DRNG,
+		.seed_life = 5,
+		.dfmul = 2,
+		.predict_en = false,
+		.iseed_en = true,
+		.pstr_en = true,
+		.df_disable = false,
+	};
+	const uint8_t ext_seed[TRNG_SEED_LEN] = {
+		0x3BU, 0xC3U, 0xEDU, 0x64U, 0xF4U, 0x80U, 0x1CU, 0xC7U,
+		0x14U, 0xCCU, 0x35U, 0xEDU, 0x57U, 0x01U, 0x2AU, 0xE4U,
+		0xBCU, 0xEFU, 0xDEU, 0xF6U, 0x7CU, 0x46U, 0xA6U, 0x34U,
+		0xC6U, 0x79U, 0xE8U, 0x91U, 0x5DU, 0xB1U, 0xDBU, 0xA7U,
+		0x49U, 0xA5U, 0xBBU, 0x4FU, 0xEDU, 0x30U, 0xB3U, 0x7BU,
+		0xA9U, 0x8BU, 0xF5U, 0x56U, 0x4DU, 0x40U, 0x18U, 0x9FU,
+	};
+	const uint8_t pers_str[TRNG_PERS_STR_LEN] = {
+		0xB2U, 0x80U, 0x7EU, 0x4CU, 0xD0U, 0xE4U, 0xE2U, 0xA9U,
+		0x2FU, 0x1FU, 0x5DU, 0xC1U, 0xA2U, 0x1FU, 0x40U, 0xFCU,
+		0x1FU, 0x24U, 0x5DU, 0x42U, 0x61U, 0x80U, 0xE6U, 0xE9U,
+		0x71U, 0x05U, 0x17U, 0x5BU, 0xAFU, 0x70U, 0x30U, 0x18U,
+		0xBCU, 0x23U, 0x18U, 0x15U, 0xCBU, 0xB8U, 0xA6U, 0x3EU,
+		0x83U, 0xB8U, 0x4AU, 0xFEU, 0x38U, 0xFCU, 0x25U, 0x87U,
+	};
+	const uint8_t expected_out[TRNG_GEN_LEN] = {
+		0x91U, 0x9AU, 0x6BU, 0x99U, 0xD5U, 0xBCU, 0x2CU, 0x11U,
+		0x5FU, 0x3AU, 0xFCU, 0x0BU, 0x0EU, 0x7BU, 0xC7U, 0x69U,
+		0x4DU, 0xE1U, 0xE5U, 0xFEU, 0x59U, 0x9EU, 0xAAU, 0x41U,
+		0xD3U, 0x48U, 0xFDU, 0x3DU, 0xD2U, 0xC4U, 0x50U, 0x1EU,
+	};
+	uint8_t out[TRNG_GEN_LEN] = { 0 };
+
+	if (!trng)
+		return TEE_ERROR_GENERIC;
+
+	memcpy(&tests.init_seed, ext_seed, sizeof(ext_seed));
+	memcpy(tests.pstr, pers_str, sizeof(pers_str));
+
+	if (trng_instantiate(trng, &tests))
+		goto error;
+
+	if (trng_generate(trng, out, sizeof(out), false))
+		goto error;
+
+	if (memcmp(out, expected_out, TRNG_GEN_LEN)) {
+		EMSG("K.A.T mismatch");
+		goto error;
+	}
+
+	if (trng_release(trng))
+		goto error;
+
+	return TEE_SUCCESS;
+error:
+	trng->status = TRNG_ERROR;
+	return TEE_ERROR_GENERIC;
+}
+
+static struct versal_trng versal_trng = {
+	.cfg.base = TRNG_BASE,
+	.cfg.len = TRNG_SIZE,
+};
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
```

Comment:
Does it need a header file to declare this interface? so that others can use this interface.

Reply:
no it is not required. the interface is already in core/include/rng_support.h (it is not driver specific)